compiler/src/iree/compiler/Codegen/LLVMCPU/LLVMCPULowerExecutableTarget.cpp

```diff
@@ -185,6 +198,8 @@ void LLVMCPULowerExecutableTargetPass::runOnOperation() {
 
       auto target = variantOp.getTarget();
       bool lowerToAVX2 = hasAVX2Feature(target);
+      bool lowerTransposeToShuffle16x16 =
+          hasAVX512fFeature(target) && hasPackMatmulLhsOp(moduleOp);
```

Comment:
I think it's worth create a `TargetFeatures` struct or similar to pass the avx2 and avx512 support to the pipelines.

Also wondering why we have to check for hasPackMatmulLhsOp here. If we just happen to have avx512 support and find a 16x16 transpose, regardless of where it's coming from, it should be lowered using the 16x16 avx512 pattern, right?

Reply:
I feel that things are disconnected between kernel configurations and these vector lowering options. We want some vector strategy when setting lowering_config, but we can't get the vector strategy when configuring the pipeline. E.g., we want to enable 16x16 transpose shuffle if we find a 16x16 transpose and we have avx512f features. It is easy to detect when we set configs for kernels, but not build the pipeline.

I'm thinking to introduce `TargetFeatures` which is inspired from your recent multi-tiling work. But it is still not clear enough to me about how do the pipeline know if there is a 16x16 transpose or not w/o looking into operations.

It is maybe a good example about what we've discussed last week. We want to set some strategies to somewhere, and ask the pipeline to extract the configurations from it.

Reply:
With https://reviews.llvm.org/D149908, there are two situations:

1. If they are 2D transpose, they will fall back to use shuffle_1d approach. It lowers it to a big 1D shuffle op.
2. If not, they become scalar loads/stores.

Reply:
What is your suggestion to move it forward? I can drop this kind of change. We can force using shuffle_16x16 in LLVMCPUVectorLowering if we find a 16x16 vector.transpose case. IMO, it's good because we still can use the strategy for 16x16 transpose even there are no avx512f features.

Reply:
No longer need to pass the option. I added the detection at LLVMCPUVectorLowering pass.